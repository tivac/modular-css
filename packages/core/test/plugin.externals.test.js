"use strict";

var path   = require("path"),

    postcss = require("postcss"),
    
    plugin  = require("../plugins/externals.js"),
    resolve = require("../lib/resolve.js").resolve,
    
    processor = postcss([ plugin ]);

describe("/plugins", function() {
    describe("/externals.js", function() {
        var start = path.resolve("./packages/core/test/specimens/start.css"),
            local = path.resolve("./packages/core/test/specimens/local.css");

        // Helper to create environment where other files are already processed
        function process(css) {
            return processor.process(css, {
                resolve,

                from  : start,
                files : {
                    // Composition source
                    [start] : {},

                    // Composition target
                    [local] : {
                        exports : {
                            fooga : [ "a", "b" ]
                        }
                    }
                }
            });
        }
        
        it("should fail to parse invalid declarations", function() {
            expect(
                () => process(`:external(red from "./local.css) {}`).css
            )
            .toThrow(/: Unclosed string/);
            
            expect(
                () => process(`:external(blue from) {} `).css
            )
            .toThrowErrorMatchingSnapshot();
        });

        it("should fail if there's no file reference", function() {
            expect(
                () => process(`:external(booga) {}`).css
            )
            .toThrow(/externals must be from another file/);
        });

        it("should fail if importing from a file that doesn't exist", function() {
            expect(
                () => process(`:external(booga from "./no.css") {}`).css
            )
            .toThrow(/Unknown external source/);
        });

        it("should fail if a non-existant selector is referenced", function() {
            expect(
                () => process(`:external(booga from "./local.css") {}`).css
            )
            .toThrow(/Invalid external reference: booga/);
        });

        it("should support importing a selector from another file", function() {
            expect(
                process(`.a :external(fooga from "./local.css") {}`).css
            )
            .toMatchSnapshot();
        });
    });
});
