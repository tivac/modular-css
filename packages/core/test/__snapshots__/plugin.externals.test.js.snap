// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/plugins /externals.js should fail to parse invalid declarations 1`] = `"Expected source but end of input found."`;

exports[`/plugins /externals.js should support importing a selector from another file 1`] = `".a .a.b {}"`;
